tests/unit/db/database_tests.py

```diff
@@ -0,0 +1,480 @@
+        self.db.create_document(data)
+        ddoc_list = self.db.list_design_documents()
+        self.assertTrue(all(x in ddoc_list for x in self.db.keys()))
+
+    def test_retrieve_design_document(self):
+        """
+        Test retrieve a specific design document
+        """
+        # Get an empty design document object that does not exist remotely
+        local_ddoc = self.db.get_design_document('_design/ddoc01')
+        self.assertEqual(local_ddoc, {'views': {}})
+
+        # Add the design document to the database
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        local_ddoc.add_view('view01', map_func)
+        local_ddoc.save()
+
+        # Get the recently created design document that now exists remotely
+        ddoc = self.db.get_design_document('_design/ddoc01')
+        self.assertEqual(ddoc, local_ddoc)
+
+    def test_retrieve_view_results(self):
+        """
+        Test retrieving Result wrapped output from a design document view
+        """
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        data = {'_id': '_design/ddoc01','views': {'view01': {"map": map_func}}}
+        self.db.create_document(data)
+        self.populate_db_with_documents()
+
+        # Test with default Result
+        rslt = self.db.get_view_result('_design/ddoc01', 'view01')
+        self.assertIsInstance(rslt, Result)
+        self.assertEqual(rslt[:1], rslt['julia000'])
+
+        #Test with custom Result
+        rslt = self.db.get_view_result(
+            '_design/ddoc01',
+            'view01',
+            descending=True,
+            reduce=False)
+        self.assertIsInstance(rslt, Result)
+        self.assertEqual(rslt[:1], rslt['julia099'])
+
+    def test_retrieve_raw_view_results(self):
+        """
+        Test retrieving raw output from a design document view
+        """
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        data = {'_id': '_design/ddoc01','views': {'view01': {"map": map_func}}}
+        self.db.create_document(data)
+        self.populate_db_with_documents()
+
+        raw_rslt = self.db.get_view_raw_result('_design/ddoc01', 'view01')
+        self.assertIsInstance(raw_rslt, dict)
+        self.assertEqual(len(raw_rslt.get('rows')), 100)
+
+    def test_all_docs(self):
+        """
+        Test the all_docs functionality
+        """
+        self.populate_db_with_documents()
+        data = self.db.all_docs(
+            limit=3,
+            keys=['julia006', 'julia024', 'julia045', 'julia099']
+        )
+        self.assertEqual(len(data.get('rows')), 3)
+        self.assertEqual(data['rows'][0]['key'], 'julia006')
+        self.assertEqual(data['rows'][1]['key'], 'julia024')
+        self.assertEqual(data['rows'][2]['key'], 'julia045')
+
+    def test_custom_result_context_manager(self):
+        """
+        Test using the database custom result context manager
+        """
+        self.populate_db_with_documents()
```

Comment:
should we test all of the different kinds of keyword option? https://github.com/cloudant/python-cloudant/blob/master/src/cloudant/result.py#L28-L43